pkg/devfile/resources.go

```diff
@@ -0,0 +1,258 @@
+		TypeMeta: generator.GetTypeMeta("Route", "route.openshift.io/v1"),
+		RouteSpecParams: generator.RouteSpecParams{
+			ServiceName: name,
+			PortNumber:  intstr.FromString(port),
+			Path:        path,
+			Secure:      secure,
+		},
+	}
+
+	return *generator.GetRoute(devfilev1.Endpoint{Annotations: annotations}, routeParams)
+}
+
+// GetResourceFromDevfile gets the deployment, service and route resources from the devfile
+func GetResourceFromDevfile(devfileObj parser.DevfileObj, deployAssociatedComponents map[string]string, name string) (appsv1.Deployment, corev1.Service, routev1.Route, error) {
+	kubernetesComponentFilter := common.DevfileOptions{
+		ComponentOptions: common.ComponentOptions{
+			ComponentType: devfilev1.KubernetesComponentType,
+		},
+	}
+	kubernetesComponents, err := devfileObj.Data.GetComponents(kubernetesComponentFilter)
+	if err != nil {
+		errMsg := fmt.Sprintf("Failed to get the kubernetes components from devfile: %v", err)
+		klog.Error(errMsg)
+		return appsv1.Deployment{}, corev1.Service{}, routev1.Route{}, err
+	}
+
+	var appendedResources parser.KubernetesResources
+	var deployment appsv1.Deployment
+	var service corev1.Service
+	var route routev1.Route
+
+	returnDeploymentOnly := true // NOTE: (TODO) once devfile samples YAML have service & route definitions, switch to false
+	// Update tests accordingly
+
+	for _, component := range kubernetesComponents {
+		if _, ok := deployAssociatedComponents[component.Name]; ok && component.Kubernetes != nil {
+			var src parser.YamlSrc
+
+			if component.Kubernetes.Inlined != "" {
+				src = parser.YamlSrc{
+					Data: []byte(component.Kubernetes.Inlined),
+				}
+			} else {
+				// If the Kubernetes YAML file does not have inline content, fall back to the dev preview way
+				// Once we have resolved the flow between frontend, backend and consume only YAML files,
+				// we can err out here instead of falling back
+
+				// Deployment
+				deploymentResource, err := GetDeployResource(name, devfileObj, common.DevfileOptions{})
+				if err != nil {
+					errMsg := fmt.Sprintf("Failed to get Deployment resource for the devfile: %v", err)
+					klog.Error(errMsg)
+					return appsv1.Deployment{}, corev1.Service{}, routev1.Route{}, err
+				}
+				appendedResources.Deployments = append(appendedResources.Deployments, deploymentResource)
+
+				dockerImagePort := devfileObj.Data.GetMetadata().Attributes.GetString("alpha.dockerimage-port", &err)
+				if err != nil {
+					errMsg := fmt.Sprintf("Failed to get the Dockerfile location from devfile metadata attribute 'alpha.build-dockerfile': %v", err)
+					klog.Error(errMsg)
+
+					return appsv1.Deployment{}, corev1.Service{}, routev1.Route{}, err
+				}
+
+				imageComponentFilter := common.DevfileOptions{
+					ComponentOptions: common.ComponentOptions{
+						ComponentType: devfilev1.ImageComponentType,
+					},
+				}
+
+				// Service
+				serviceResource, err := GetService(devfileObj, imageComponentFilter, dockerImagePort)
```

Comment:
why want to pass in image component filter as an arg? can this be defined within the function itself?